packages/controller/src/lib/upgradeManager.ts

```diff
@@ -0,0 +1,299 @@
+        this.response.success = res.success;
+
+        if (!res.success) {
+            throw new Error(`Could not install js-controller@${this.version}`);
+        }
+    }
+
+    /**
+     * Starts the web server for admin communication either secure or insecure
+     *
+     * @param params Web server configuration
+     */
+    startWebServer(params: WebServerParameters): void {
+        const { useHttps } = params;
+        if (useHttps) {
+            this.startSecureWebServer(params);
+        } else {
+            this.startInsecureWebServer(params);
+        }
+    }
+
+    /**
+     * This function is called when the webserver receives a message
+     *
+     * @param req received message
+     * @param res server response
+     * @param server https or https server instance
+     */
+    webServerCallback(req: http.IncomingMessage, res: http.ServerResponse, server: http.Server | https.Server): void {
+        res.writeHead(200);
+        res.end(JSON.stringify(this.response));
+
+        if (!this.response.running) {
+            console.log('Final information delivered');
+
+            server.close(async () => {
+                await this.startController();
+                console.log('Successfully started js-controller');
+
+                process.exit();
+            });
+        }
+    }
+
+    /**
+     * Start an insecure web server for admin communication
+     *
+     * @param params Web server configuration
+     */
+    startInsecureWebServer(params: InsecureWebServerParameters): void {
+        const { port } = params;
+
+        const server = http.createServer((req, res) => {
+            this.webServerCallback(req, res, server);
+        });
+
+        server.listen(port, () => {
+            console.log(`Server is running on http://localhost:${port}`);
+        });
+    }
+
+    /**
+     * Start a secure web server for admin communication
+     *
+     * @param params Web server configuration
+     */
+    startSecureWebServer(params: SecureWebServerParameters): void {
+        const { port, certPublic, certPrivate } = params;
+
+        const server = https.createServer({ key: certPrivate, cert: certPublic }, (req, res) => {
+            this.webServerCallback(req, res, server);
+        });
+
+        server.listen(port, () => {
+            console.log(`Server is running on http://localhost:${port}`);
+        });
+    }
+
+    /**
+     * Get certificates from the DB
+     *
+     * @param params certificate information
+     */
+    async getCertificates(params: GetCertificatesParams): Promise<Certificates> {
+        const { objects, certPublicName, certPrivateName } = params;
+
+        const obj = await objects.getObjectAsync('system.certificates');
+
+        if (!obj) {
+            throw new Error('No certificates found');
+        }
+
+        const certs = obj.native.certificates;
+
+        return { certPrivate: certs[certPrivateName], certPublic: certs[certPublicName] };
```

Comment:
does that work with acme too? :-)

Reply:
tbh: I hope so, the certificate names in the interal object should stay the same